src/Functions.tsx

```diff
@@ -150,9 +155,43 @@ const updateExistingRows = (
 
 const createCombinedRow = (
   averageCoordinates: Point,
-  rows: Types.row[]
+  rows: Types.row[],
+  index: number
 ): Types.combinedRow => {
-  return { averageCoordinates: averageCoordinates, rows: rows };
+  return {
+    averageCoordinates: averageCoordinates,
+    rows: rows,
+    index: index,
+    isMarkerVisible: getCombinedRowVisibility(index),
+  };
+};
+
+export const getCombinedRowVisibility = (index: number): Types.visibility => {
+  const elementId = getMarkerId(index);
+  const svgMarker = document.getElementById(elementId);
+  const isMarkerVisible = elementIsInViewport(svgMarker);
+  return isMarkerVisible;
+};
+
+export const getMarkerId = (index: number): string =>
+  Config.elementIds.Marker(index);
+
+const elementIsInViewport = (element: HTMLElement | null): Types.visibility => {
+  const rectangle = element?.getBoundingClientRect();
+  const container = document.getElementById("container");
+  const elementNotValid = !element || 1 !== element.nodeType;
+  const noSuchRectangle = rectangle === undefined;
+  const noSuchContainer = container === null;
```

Comment:
Mixing `null` and `undefined` is not ideal. Can you review your codebase, find all instances of `null` and change them to `undefined`? Then from here on out, only use `undefined`. This makes your checks a lot easier.

Reply:
`null`: displayDebuggingFeatures and elementIsInViewport
`undefined`: Config.tsx, elementIsInViewPort, and Types.tsx

It looks like replacing null in elementIsInViewport is safe